cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ImpersonationHandler.java

```diff
@@ -81,7 +82,13 @@ public void getCredentials(HttpRequest request, HttpResponder responder) throws
     }
     ImpersonationRequest impersonationRequest = GSON.fromJson(requestContent, ImpersonationRequest.class);
     LOG.debug("Fetching credentials for {}", impersonationRequest);
-    UGIWithPrincipal ugiWithPrincipal = ugiProvider.getConfiguredUGI(impersonationRequest);
+    UGIWithPrincipal ugiWithPrincipal;
+    try {
+      ugiWithPrincipal = ugiProvider.getConfiguredUGI(impersonationRequest);
+    } catch (IOException e) {
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, e.getMessage());
```

Comment:
throw to let the `HttpExceptionHandler` handle it instead of using the responder directly

Reply:
See my message here for why I did this: https://github.com/caskdata/cdap/pull/9552#issuecomment-326448141

Reply:
I don't understand how they are related. You can throw an exception that implements `HttpErrorStatusProvider` to control both status code and the message. Throwing exception clear the possibility of missing the `return` statement.

Reply:
It's related because the HttpExceptionHandler is responding with the root cause's exception message, which is not a useful message.

Reply:
You can throw one without a cause, right?

Reply:
e.g. `throw new IOException(e.getMessage());`

Reply:
I want to keep the stack trace, since we're not sure what all the different errors could be encountered.
If we hit a different type of error, the stack trace might be useful.
I'd rather go the route of throwing an `HttpErrorStatusProvider`, or just leave this `catch` in the http handler.

Reply:
But you are not responding with the stacktrace

Reply:
In other usages, this code path is not called from a http handler; its useful when logged there.

Reply:
?? This is using a http responder. Also, what do mean by logged there?? The exception is not thrown out.

Reply:
But its also used in other places by impersonator, which can be outside an http handler, such as during upgrade tool or many other places.

Reply:
You mean this method is called from impersonator/upgrade directly?? What http responder it provides? Also, even so, how is the stack trace being used?? I still don't get it.

Reply:
It doesn't call the handler's method directly, but it calls UGIProvider method directly.
The stacktrace will be logged if there is an error encountered by the UGIProvider.

Reply:
So some other classes calls `UGIProvider.xxx` directly.. but how does that related to this handler class??

Reply:
Ah, I thought you meant to throw an IOException differently in the UGIProvider implementation. You mean just throw an IOException with a simple message here. Will update.

Reply:
I tried this out, but the downside to this approach is that then the exception will also get logged on the server side.
I think its fine just to catch the IOException and respond with the exception message, instead of also logging the exception on the server side.
In order to avoid it then, we'd need to introduce a custom `HttpErrorStatusProvider` class, but it seems simple just to use catch + respond + return.